### PR TITLE
Extend TrackBase and Track to include time and velocity

### PR DESCRIPTION
**Pull request by lgray**
Add time, velocity, and corresponding diagonal covariance terms to reco::TrackBase.
These (in addition to #24419) are the minimal changes needed for MTD development.

This is purely technical, getting it in place so that we can ease development.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

File: DataFormats/TrackReco/interface/TrackBase.h [hunk 1]
```diff
@@ -683,6 +710,18 @@ inline const TrackBase::Point & TrackBase::referencePoint() const
...
+// Time at the reference point on the track
+inline double TrackBase::t0() const
+{
+    return t0_;
+}
+
+// Velocity at the reference point on the track
+inline double TrackBase::beta() const
+{
+    return beta_;
```
Comment (slava77) [inline DataFormats/TrackReco/interface/TrackBase.h]: Is this an absolute velocity, or is it normalized to c? Please clarify in the comments.

File: DataFormats/TrackReco/interface/TrackBase.h [hunk 2]
```diff
@@ -816,6 +855,18 @@ inline double TrackBase::dzError() const
...
+// error on t0
+inline double TrackBase::t0Error() const
+{
+    return std::sqrt(covt0t0_);
+}
+
+// error on beta
+inline double TrackBase::betaError() const
+{
+    return std::sqrt(covbetabeta_);
```
Comment (slava77) [inline DataFormats/TrackReco/interface/TrackBase.h]: there should also be a way to get these without sqrt.
Perhaps covarianceT0T0 and covarianceBetaBeta.
BTW, are the two strictly related by a track-dependent constant or is there a T0Beta term?

File: DataFormats/TrackReco/src/Track.cc [hunk 3]
```diff
@@ -4,8 +4,11 @@ using namespace reco;
 
 Track::Track(double chi2, double ndof, const Point &vertex, const Vector &momentum,
              int charge, const CovarianceMatrix &cov, TrackAlgorithm algo,
-             TrackQuality quality) :
-    TrackBase(chi2, ndof, vertex, momentum, charge, cov, algo, quality)
+             TrackQuality quality, double t0, double beta, 
+	     double covt0t0, double covbetabeta) :
+  TrackBase(chi2, ndof, vertex, momentum, charge, cov, algo, quality,
+	    0,0, // nloops and stop reason
+	    t0,beta,covt0t0,covbetabeta)
```
Comment (slava77) [inline DataFormats/TrackReco/src/Track.cc]: indentation

File: DataFormats/TrackReco/src/TrackBase.cc [hunk 4]
```diff
@@ -89,10 +89,15 @@ TrackBase::TrackBase() :
 
 TrackBase::TrackBase(double chi2, double ndof, const Point &vertex, const Vector &momentum,
                      int charge, const CovarianceMatrix &cov, TrackAlgorithm algorithm,
-                     TrackQuality quality, signed char nloops, uint8_t stopReason):
+                     TrackQuality quality, signed char nloops, uint8_t stopReason,
+		     double t0, double beta, double covt0t0, double covbetabeta):
```
Comment (slava77) [inline DataFormats/TrackReco/src/TrackBase.cc]: indentation

File: DataFormats/TrackReco/interface/TrackBase.h [hunk 5]
```diff
@@ -407,14 +424,24 @@ class TrackBase
...
 
+    /// time at the reference point on track
+    float t0_;
+
     /// momentum vector at innermost point
     Vector momentum_;
+    
+    /// norm of the particle velocity at innermost point on track
+    /// can multiply by momentum_.Unit() to get velocity vector
+    float beta_;
```
Comment (slava77) [inline DataFormats/TrackReco/interface/TrackBase.h]: please define default values here or in the default constructor.
I can't recall now if they also should be specified in the ioread rules (it's easy to check by reading an old file and checking what's in covt0t0_ )

**Comment by slava77** [comment]
@lgray 
please clarify on the plans for updating the PR to follow up on the code review.
Thank you.

**Comment by lgray** [comment]
I'll update this by the end of today.

**Comment by slava77** [comment]
> I'll update this by the end of today.

@lgray 
 please clarify on the plans for updating the PR to follow up on the code review.
thank you.

**Comment by lgray** [comment]
@slava77 working on it now!

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24465	4419bc4
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no [relevant] differences
- local tests confirm that the default values are correctly propagated when reading old data files

**Comment by fabiocos** [comment]
please test workflow 22434.0

**Comment by fabiocos** [comment]
+1

**Pull request merged**